### PR TITLE
Performance: slow edit specialists page

### LINKED ISSUE #2321: [Performance]: Edit Specialists page is very slow to load
### Description

The Edit Specialists page takes a long time to load when the system contains a lot of specialists. For MHI, we have 10384 specialists and it takes 10+ seconds for the page to load. 

Our administrative staff need to access this page multiple times a day so the wait time is a hindrance. 

### Steps to Reproduce

0. Create a lot of entries in the `professionalSpecialists` table
1. From the main schedule page, navigate to Consultations > Edit Specialists Listing > Edit Specialists

Alternatively, navigate to Consultations > Edit Specialists Listing > Show All Services > click on a service

### Expected Performance

Page should load quickly

### Actual Performance

Page takes 10+ seconds to load

### Module/Area

Other

### Environment

10.3k specialists

### Metrics/Logs

```shell

```

### Checklist

- [x] This happens consistently (not intermittent)
- [ ] I have tested with minimal data to rule out data volume issues

### PR DESCRIPTION
**Pull request by LiamStanziani**
## Summary                                                                                                                                                                                                                                
  Decouple specialist list data from inline JSP rendering into a separate                                                                                                                                                                   
  fetch() POST endpoint with client-side batch rendering.  

Fixes #2321                                                                                                                                                                                  
                                                                                                                                                                                                                                            
  ## Problem                                                                                                                                                                                                                                
  The Edit Specialists and Display Service pages embed all specialist data                                                                                                                                                                  
  directly in the HTML response via server-side `<% for %>` loops. With                                                                                                                                                                     
  10k+ specialist records this causes:                                                                                                                                                                                                      
                                                                                                                                                                                                                                            
  - ~226MB transient memory allocation per request (DTO list + JSON string                                                                                                                                                                  
    + JSP output buffer all held simultaneously)                                                                                                                                                                                            
  - 1,705KB HTML response containing the full specialist table                                                                                                                                                                              
  - ~4 second browser (local environment) DOM parsing delay before the page becomes interactive                                                                                                                                                                                                                                                                                                                                      
                                                                                                                                                                                                                                            
  ## Solution                                                                                                                                                                                                                               
                                                                                                                                                                                                                                            
  **DTO Projection (server):**                                                                                                                                                                                                              
  - JPQL constructor projection (`SELECT NEW SpecialistListDTO(...)`)                                                                                                                                                                       
    fetches only the 7 columns needed for list display, avoiding full                                                                                                                                                                       
    entity hydration of heavy text fields (annotation, eDataOscarKey, etc.)                                                                                                                                                                 
                                                                                                                                                                                                                                          
  **Separate JSON endpoint (server):**                                                                                                                                                                                                      
  - New `SpecialistList2Action` serves specialist data as JSON via POST,                                                                                                                                                                    
    following the same pattern as `ConsultationLookup2Action` and                                                                                                                                                                           
    `TicklerList2Action`                                                                                                                                                                                                                    
  - Jackson ArrayNode builds the response, written directly to                                                                                                                                                                              
    `response.getWriter()` — decouples data payload from page shell                                                                                                                                                                         
                                                                                                                                                                                                                                            
  **Batch rendering (client):**                                                                                                                                                                                                             
  - JSP pages load a lightweight shell (~9KB) with spinner                                                                                                                                                                                  
  - `fetch()` POST retrieves JSON from the new endpoint                                                                                                                                                                                     
  - `requestAnimationFrame` + `document.createDocumentFragment()` renders                                                                                                                                                                   
    rows in batches of 1,000 for instant perceived load

## Summary by Sourcery

Introduce a JSON-based specialist list endpoint and client-side rendering to decouple large specialist datasets from JSP pages and improve performance on specialist configuration screens.

New Features:
- Add SpecialistList2Action JSON endpoints to serve specialist list data for Edit Specialists and Display Service pages.
- Introduce SpecialistListDTO as a lightweight projection for specialist list views.

Enhancements:
- Refactor EditSpecialists.jsp and DisplayService.jsp to fetch specialist data asynchronously and render rows client-side using a shared renderer script.
- Optimize specialist data access in ProfessionalSpecialistDao with a DTO-based query that selects only columns needed for list display.
- Simplify EctConDisplayServiceUtil by removing specialist list construction responsibilities.

Tests:
- Add integration tests verifying ProfessionalSpecialistDao DTO projection behavior, ordering, and null-handling.

<!-- This is an auto-generated description by cubic. -->
---
## Summary by cubic
Move specialist data out of JSP into CSRF-protected JSON endpoints with client-side batch rendering to speed up Edit Specialists and Display Service; cuts payload and server memory for much faster interactivity. Fixes #2321.

- **Refactors**
  - Add `SpecialistList2Action` with POST-only endpoints `getSpecialists` and `getSpecialistsForService`, CsrfGuard header support, `_con` read privilege check, and 400/405 error handling; mapped in `struts.xml`.
  - Add `ProfessionalSpecialistDao#findAllListDTOs()` using JPQL constructor projection to `SpecialistListDTO` (id, names, letters, address, phone, fax); include integration tests.
  - Add `specialistListRenderer.js` for batched `requestAnimationFrame` rendering and CSRF header; update `EditSpecialists.jsp` and `DisplayService.jsp` to fetch JSON and render checkboxes/links, keeping the spinner until the first batch renders.
  - Remove specialist vector building and inline JSP loops from `EctConDisplayServiceUtil` and drop related calls (e.g., from `ConsultationFormRequest.jsp`).

<sup>Written for commit a0d1175a91b192079ab19517a1a736996866fa74. Summary will update on new commits.</sup>

<!-- End of auto-generated description by cubic. -->

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **New Features**
  * Added API endpoint for retrieving specialist information, including support for filtering specialists by service and checking assignment status.

* **Refactor**
  * Improved specialist list display performance by switching from server-side rendering to dynamic client-side loading with batched data rendering.
  * Streamlined specialist data initialization in consultation request workflows.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

### CODE REVIEW

**Comment by LiamStanziani** [comment]
@sourceryai review

**Comment by LiamStanziani** [comment]
@coderabbitai review

File: src/main/java/ca/openosp/openo/encounter/oscarConsultationRequest/config/pageUtil/SpecialistList2Action.java [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+        String method = request.getParameter("method");
+
+        if ("getSpecialists".equals(method)) {
+            return getSpecialists();
+        } else if ("getSpecialistsForService".equals(method)) {
+            return getSpecialistsForService();
+        }
+
+        MiscUtils.getLogger().warn("SpecialistList2Action: invalid method parameter: " + method);
```
[the remark this answers is not in the document]
Comment (LiamStanziani) [reply src/main/java/ca/openosp/openo/encounter/oscarConsultationRequest/config/pageUtil/SpecialistList2Action.java]: As explained, fixed in the recent commit

File: src/main/java/ca/openosp/openo/commn/dao/ProfessionalSpecialistDaoImpl.java [hunk 2]
```diff
@@ -55,6 +56,22 @@ public List<ProfessionalSpecialist> findAll() {
         return (results);
     }
 
+    @Override
+    public List<SpecialistListDTO> findAllListDTOs() {
+        Query query = entityManager.createQuery(
```
[the remark this answers is not in the document]
Comment (LiamStanziani) [reply src/main/java/ca/openosp/openo/commn/dao/ProfessionalSpecialistDaoImpl.java]: As explained, fixed in the recent commit

File: src/main/webapp/oscarEncounter/oscarConsultationRequest/config/EditSpecialists.jsp [hunk 3]
```diff
@@ -178,5 +136,71 @@
...
+                    tr.appendChild(nameTd);
+
+                    tr.appendChild(createCell(s[2]));
+                    tr.appendChild(createCell(s[3]));
+                    tr.appendChild(createCell(s[4]));
+
+                    fragment.appendChild(tr);
+                }
+                tbody.appendChild(fragment);
+                if (idx === Math.min(BATCH_SIZE, data.length)) {
```
[the remark this answers is not in the document]
Comment (LiamStanziani) [reply src/main/webapp/oscarEncounter/oscarConsultationRequest/config/EditSpecialists.jsp]: As explained, fixed in the recent commit

File: src/main/java/ca/openosp/openo/encounter/oscarConsultationRequest/config/pageUtil/SpecialistList2Action.java [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+     *
+     * @param dto SpecialistListDTO the specialist data
+     * @return String the formatted display name
+     */
+    private static String buildDisplayName(SpecialistListDTO dto) {
+        String lName = StringUtils.defaultString(dto.getLastName());
+        String fName = StringUtils.defaultString(dto.getFirstName());
+        String proLetters = dto.getProfessionalLetters();
+        return lName + " " + fName + (proLetters == null ? "" : " " + proLetters);
+    }
```
[the remark this answers is not in the document]
Comment (LiamStanziani) [reply src/main/java/ca/openosp/openo/encounter/oscarConsultationRequest/config/pageUtil/SpecialistList2Action.java:204]: Fixed in the recent commit

File: src/main/webapp/oscarEncounter/oscarConsultationRequest/config/EditSpecialists.jsp [hunk 5]
```diff
@@ -178,5 +136,71 @@
...
+                }
+                if (idx < data.length) {
+                    requestAnimationFrame(function() { renderBatch(data, idx); });
+                }
+            }
+
+            fetch("<%= request.getContextPath() %>/oscarEncounter/SpecialistList.do?method=getSpecialists", {
+                method: "POST"
+            })
+            .then(function(resp) { return resp.json(); })
```
[the remark this answers is not in the document]
Comment (LiamStanziani) [reply src/main/webapp/oscarEncounter/oscarConsultationRequest/config/EditSpecialists.jsp]: Fixed in the recent commit

File: src/main/webapp/oscarEncounter/oscarConsultationRequest/config/EditSpecialists.jsp [hunk 6]
```diff
@@ -178,5 +136,71 @@
             </tr>
         </table>
     </div>
+    <script>
+        (function() {
+            var BATCH_SIZE = 1000;
+            var tbody = document.getElementById("specialistBody");
+            var editUrlPrefix = "<%= request.getContextPath() %>/oscarEncounter/EditSpecialists.do?specId=";
+
```
[the remark this answers is not in the document]
Comment (LiamStanziani) [reply src/main/webapp/oscarEncounter/oscarConsultationRequest/config/EditSpecialists.jsp]: Fixed in the recent commit

File: src/main/webapp/oscarEncounter/oscarConsultationRequest/config/DisplayService.jsp [hunk 7]
```diff
@@ -189,5 +148,65 @@
...
+                }
+                if (idx < data.length) {
+                    requestAnimationFrame(function() { renderBatch(data, idx); });
+                }
+            }
+
+            fetch("<%= request.getContextPath() %>/oscarEncounter/SpecialistList.do?method=getSpecialistsForService&serviceId=<%=serviceId %>", {
+                method: "POST"
+            })
+            .then(function(resp) { return resp.json(); })
```
[the remark this answers is not in the document]
Comment (LiamStanziani) [reply src/main/webapp/oscarEncounter/oscarConsultationRequest/config/DisplayService.jsp]: Fixed in the recent commit

File: src/main/java/ca/openosp/openo/encounter/oscarConsultationRequest/config/pageUtil/SpecialistList2Action.java [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+        String method = request.getParameter("method");
+
+        if ("getSpecialists".equals(method)) {
+            return getSpecialists();
+        } else if ("getSpecialistsForService".equals(method)) {
+            return getSpecialistsForService();
+        }
+
+        MiscUtils.getLogger().warn("SpecialistList2Action: invalid method parameter: " + method);
+        return NONE;
```
[the remark this answers is not in the document]
Comment (LiamStanziani) [reply src/main/java/ca/openosp/openo/encounter/oscarConsultationRequest/config/pageUtil/SpecialistList2Action.java]: Fixed in the recent commit

File: src/main/webapp/oscarEncounter/oscarConsultationRequest/config/DisplayService.jsp [hunk 9]
```diff
@@ -189,5 +148,65 @@
...
+                tbody.appendChild(fragment);
+                if (idx === Math.min(BATCH_SIZE, data.length)) {
+                    HideSpin();
+                }
+                if (idx < data.length) {
+                    requestAnimationFrame(function() { renderBatch(data, idx); });
+                }
+            }
+
+            fetch("<%= request.getContextPath() %>/oscarEncounter/SpecialistList.do?method=getSpecialistsForService&serviceId=<%=serviceId %>", {
```
[the remark this answers is not in the document]
Comment (LiamStanziani) [reply src/main/webapp/oscarEncounter/oscarConsultationRequest/config/DisplayService.jsp]: Fixed in the recent commit

File: src/main/java/ca/openosp/openo/consultation/dto/SpecialistListDTO.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+        this.faxNumber = faxNumber;
+    }
+
+    public Integer getId() { return id; }
+    public String getFirstName() { return firstName; }
+    public String getLastName() { return lastName; }
+    public String getProfessionalLetters() { return professionalLetters; }
+    public String getStreetAddress() { return streetAddress; }
+    public String getPhoneNumber() { return phoneNumber; }
+    public String getFaxNumber() { return faxNumber; }
```
[the remark this answers is not in the document]
Comment (LiamStanziani) [reply src/main/java/ca/openosp/openo/consultation/dto/SpecialistListDTO.java:70]: Will be fixed in the next commit

File: src/main/java/ca/openosp/openo/encounter/oscarConsultationRequest/config/pageUtil/SpecialistList2Action.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (LiamStanziani) [reply src/main/java/ca/openosp/openo/encounter/oscarConsultationRequest/config/pageUtil/SpecialistList2Action.java]: Will be removing this in the next commit

File: src/main/webapp/oscarEncounter/oscarConsultationRequest/config/DisplayService.jsp [hunk 11]
```diff
@@ -189,5 +148,65 @@
...
+                method: "POST"
+            })
```
[the remark this answers is not in the document]
Comment (LiamStanziani) [reply src/main/webapp/oscarEncounter/oscarConsultationRequest/config/DisplayService.jsp]: Fixed in the recent commit

File: src/main/java/ca/openosp/openo/commn/dao/ProfessionalSpecialistDaoImpl.java [hunk 12]
```diff
@@ -55,6 +56,22 @@ public List<ProfessionalSpecialist> findAll() {
...
+                "SELECT NEW ca.openosp.openo.consultation.dto.SpecialistListDTO("
+                        + "x.id, x.firstName, x.lastName, x.professionalLetters, "
```
[the remark this answers is not in the document]
Comment (LiamStanziani) [reply src/main/java/ca/openosp/openo/commn/dao/ProfessionalSpecialistDaoImpl.java:64]: Will be added in the next commit

**Comment by LiamStanziani** [comment]
Not fully ready as I want to do some more testing with recent changes but want to see another AI review pass through when setting it to ready for review

File: src/main/webapp/oscarEncounter/oscarConsultationRequest/config/DisplayService.jsp [hunk 13]
```diff
@@ -70,18 +70,12 @@
         </script>
 
         <link rel="stylesheet" type="text/css" href="<%= request.getContextPath() %>/css/encounterStyles.css">
+        <script type="text/javascript" src="<%= request.getContextPath() %>/js/specialistListRenderer.js"></script>
```
[the remark this answers is not in the document]
Comment (LiamStanziani) [reply src/main/webapp/oscarEncounter/oscarConsultationRequest/config/DisplayService.jsp:73]: Forgot to commit it... fixed in the latest commit

[the remark this answers is not in the document]
Comment (LiamStanziani) [reply src/main/webapp/oscarEncounter/oscarConsultationRequest/config/DisplayService.jsp:73]: Forgot to commit it... fixed in the latest commit

**Comment by LiamStanziani** [comment]
Will need to review the remaining comments and do some testing.

File: src/main/java/ca/openosp/openo/encounter/oscarConsultationRequest/config/pageUtil/SpecialistList2Action.java [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+        } else if ("getSpecialistsForService".equals(method)) {
+            return getSpecialistsForService();
+        }
+
+        try {
+            response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Invalid or missing method parameter");
+        } catch (IOException e) {
+            MiscUtils.getLogger().error("Error sending 400 response", e);
+        }
+        return null;
```
[the remark this answers is not in the document]
Comment (LiamStanziani) [reply src/main/java/ca/openosp/openo/encounter/oscarConsultationRequest/config/pageUtil/SpecialistList2Action.java:89]: Will be updated in the next commit

File: src/main/java/ca/openosp/openo/encounter/oscarConsultationRequest/config/pageUtil/SpecialistList2Action.java [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+    }
+
+    /**
+     * Returns all specialists as a JSON array for the EditSpecialists page.
+     *
+     * <p>Each row is {@code [id, "name", "address", "phone", "fax"]}.</p>
+     *
+     * @return null (response written directly to output stream)
+     */
+    private String getSpecialists() {
```
[the remark this answers is not in the document]
Comment (LiamStanziani) [reply src/main/java/ca/openosp/openo/encounter/oscarConsultationRequest/config/pageUtil/SpecialistList2Action.java:99]: Will be updated in the next commit

File: src/main/java/ca/openosp/openo/encounter/oscarConsultationRequest/config/pageUtil/SpecialistList2Action.java [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+     *
+     * @return String action result (null when response is written directly)
+     */
+    @Override
+    public String execute() {
+        if (!securityInfoManager.hasPrivilege(LoggedInInfo.getLoggedInInfoFromSession(request), "_con", "r", null)) {
+            throw new SecurityException("missing required security object (_con)");
+        }
+
+        String method = request.getParameter("method");
```
[the remark this answers is not in the document]
Comment (LiamStanziani) [reply src/main/java/ca/openosp/openo/encounter/oscarConsultationRequest/config/pageUtil/SpecialistList2Action.java:76]: Will be updated in the next commit

File: src/main/java/ca/openosp/openo/encounter/oscarConsultationRequest/config/pageUtil/SpecialistList2Action.java [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+    /**
+     * Main dispatcher that routes to the appropriate method based on 'method' parameter.
+     *
+     * @return String action result (null when response is written directly)
+     */
+    @Override
+    public String execute() {
+        if (!securityInfoManager.hasPrivilege(LoggedInInfo.getLoggedInInfoFromSession(request), "_con", "r", null)) {
+            throw new SecurityException("missing required security object (_con)");
+        }
```
[the remark this answers is not in the document]
Comment (LiamStanziani) [reply src/main/java/ca/openosp/openo/encounter/oscarConsultationRequest/config/pageUtil/SpecialistList2Action.java:74]: This is consistent with the other areas in consultation, will leave for now

File: src/test-modern/java/ca/openosp/openo/consultation/dao/ProfessionalSpecialistDaoDTOIntegrationTest.java [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+@Tag("read")
+@Tag("dto")
+@Transactional
+public class ProfessionalSpecialistDaoDTOIntegrationTest extends OpenOTestBase {
+
+    @Autowired
+    private ProfessionalSpecialistDao professionalSpecialistDao;
+
+    @PersistenceContext(unitName = "entityManagerFactory")
+    private EntityManager entityManager;
```
[the remark this answers is not in the document]
Comment (LiamStanziani) [reply src/test-modern/java/ca/openosp/openo/consultation/dao/ProfessionalSpecialistDaoDTOIntegrationTest.java:44]: Not causing issues currently

**Comment by LiamStanziani** [comment]
Good for review at this point

**Comment by lacarmen** [comment]
Tested in our QA server. Ready for merge after Deval's review.

**Comment by D3V41** [comment]
@LiamStanziani Reviewed the PR, looks good. Do you think we should paginate the results at the SQL query level as well?

**Comment by LiamStanziani** [comment]
@D3V41 We could and I was thinking about it, the reason I didn't make that change was that it would change the behaviour of how it works (loads all at the same time with no pages). But it could be updated to that if wanted

**Comment by LiamStanziani** [comment]
At the moment I don't think its needed to fix the speed, but it could be tracked for a later improvement, which will probably need to be discussed since it will be a functional change in the pages behaviour.

**Comment by D3V41** [comment]
@LiamStanziani I don’t mind it being added separately, but I am not suggesting adding pages in the JSP page. Instead, I am proposing that we fetch all 10k records from the database in batches: fetch 1000 records from the database → show them to the user, then fetch another 1000 records → append them to the same list below, and continue this until all results are loaded. The user will not notice any difference.

But I agree with you; it will require approval before proceeding.

**Comment by D3V41** [comment]
The current implementation looks good.

**Review by D3V41 — APPROVED** [review]
